lib/interfaces/common.js

```diff
@@ -145,7 +145,16 @@ module.exports = function (suites, context, mocha) {
           throw createUnsupportedError('Pending test forbidden');
         }
         if (typeof opts.fn === 'function') {
-          opts.fn.call(suite);
+          const suiteResult = opts.fn.call(suite);
+          if (suiteResult instanceof Promise) {
```

Comment:
Going back to https://github.com/mochajs/mocha/pull/3550/files#diff-28d67f27fa361db19b8a3aa9181d89c5f0d5ae1089167c55f0341675386e3a28R115, I think we no longer need to prioritize supporting CoffeeScript users. The implicit returns that caused #3550 to be reverted in #3759 are not a good idea for the exact reason we saw here: that they unintentionally return values even when those values shouldn't be returned!

```suggestion
          if (suiteResult !== undefined) {
```

https://typescript-eslint.io/rules/no-confusing-void-expression is the right direction to go IMO.

Reply:
I agree that this will be better. And agreed that I wouldn't prioritize CoffeeScript. 

I originally tried to keep the PR minimal (wrt to potential ecosystem impact) as I'm not a maintainer, but I'm happier with your version.

Reply:
@alcuadrado is this still pending? What's up?

Reply:
Ah, I see https://github.com/mochajs/mocha/pull/4921#issuecomment-2209648248. I'd personally rather we make the breaking change & have this take longer to land. cc @mochajs/maintenance-crew for more opinions.

We've been talking about moving to new major versions soon.